src/core/lib/channel/connected_channel.cc

```diff
@@ -203,39 +242,606 @@ static void connected_channel_get_channel_info(
+                server_trailing_metadata_->DebugString().c_str(),
+                ActiveOpsString().c_str());
+      }
+      finished_ = true;
+      return ServerMetadataHandle(std::move(server_trailing_metadata_));
+    }
+    if (absl::holds_alternative<Idle>(recv_message_state_)) {
+      if (grpc_call_trace.enabled()) {
+        gpr_log(GPR_INFO, "%sPollConnectedChannel: requesting message",
+                Activity::current()->DebugTag().c_str());
+      }
+      push_recv_message();
+    }
+    if (auto* push = absl::get_if<PipeSender<MessageHandle>::PushType>(
+            &recv_message_state_)) {
+      auto r = (*push)();
+      if (bool* result = absl::get_if<bool>(&r)) {
+        if (*result) {
+          if (!finished_) {
+            if (grpc_call_trace.enabled()) {
+              gpr_log(GPR_INFO,
+                      "%sPollConnectedChannel: pushed message; requesting next",
+                      Activity::current()->DebugTag().c_str());
+            }
+            push_recv_message();
+          } else {
+            if (grpc_call_trace.enabled()) {
+              gpr_log(GPR_INFO,
+                      "%sPollConnectedChannel: pushed message and finished; "
+                      "marking closed",
+                      Activity::current()->DebugTag().c_str());
+            }
+            recv_message_state_ = Closed{};
+          }
+        } else {
+          if (grpc_call_trace.enabled()) {
+            gpr_log(GPR_INFO,
+                    "%sPollConnectedChannel: failed to push message; marking "
+                    "closed",
+                    Activity::current()->DebugTag().c_str());
+          }
+          recv_message_state_ = Closed{};
+        }
+      }
+    }
+    return Pending{};
+  }
+
+  void RecvInitialMetadataReady(grpc_error_handle error) {
+    GPR_ASSERT(error == GRPC_ERROR_NONE);
+    {
+      MutexLock lock(&mu_);
+      queued_initial_metadata_ = true;
+      initial_metadata_waker_.Wakeup();
+    }
+    Unref("initial_metadata_ready");
+  }
+
+  void RecvTrailingMetadataReady(grpc_error_handle error) {
+    GPR_ASSERT(error == GRPC_ERROR_NONE);
+    {
+      MutexLock lock(&mu_);
+      queued_trailing_metadata_ = true;
+      trailing_metadata_waker_.Wakeup();
+    }
+    Unref("trailing_metadata_ready");
+  }
+
+  void MetadataBatchDone(grpc_error_handle error) {
+    GPR_ASSERT(error == GRPC_ERROR_NONE);
+    Unref("metadata_batch_done");
+  }
+
+  void SendMessageBatchDone(grpc_error_handle error) {
+    {
+      MutexLock lock(&mu_);
+      if (error != GRPC_ERROR_NONE) {
+        send_message_state_ = Closed{};
+      }
+      if (!absl::holds_alternative<Closed>(send_message_state_)) {
+        send_message_state_ = Idle{};
+      }
+      send_message_waker_.Wakeup();
+    }
+    Unref("send_message");
+  }
+
+  void RecvMessageBatchDone(grpc_error_handle error) {
+    {
+      MutexLock lock(&mu_);
+      if (error != GRPC_ERROR_NONE) {
```

Comment:
Seems like this should close the pipe back up to the surface.

Reply:
Doesn't look like this was resolved?  Or did some of the pipe changes somehow make it so that we're closing the pipe here?

Reply:
I think we're ok in this case: if we get an error back here, we're also going to close the stream in the transport and get an error back in the promise which will close everything out and return an error to the surface.

Reply:
Ah, okay.  Can you please add a comment to that effect?